File: app/src/main/java/cl/figonzal/evaluatool/evalua1/Evalua1Activity.java
/*-----------------------------------------------------------------------------
                                                                              -
 This file is subject to the terms and conditions defined in                  -
 file 'LICENSE', which is part of this source code package                    -
                                                                              -
 Autor: Felipe González                                                       -
 Email: felipe.gonzalezalarcon94@gmail.com                                    -
                                                                              -
 Copyright (c) 2020                                                           -
                                                                              -
 Last modified 02-07-20 18:11                                                 -
 -----------------------------------------------------------------------------*/

package cl.figonzal.evaluatool.evalua1;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.firebase.crashlytics.FirebaseCrashlytics;

import java.util.ArrayList;
import java.util.List;

import cl.figonzal.evaluatool.R;
import cl.figonzal.evaluatool.adapter.EvaluaAdapter;
import cl.figonzal.evaluatool.evalua1.modulo1.MemoriaAtencionE1M1;
import cl.figonzal.evaluatool.evalua1.modulo2.ClasificacionesE1M2;
import cl.figonzal.evaluatool.evalua1.modulo2.OrganizacionPerceptivaE1M2;
import cl.figonzal.evaluatool.evalua1.modulo2.SeriesE1M2;
import cl.figonzal.evaluatool.evalua1.modulo2.ValoracionGlobalBasesE1M2;
import cl.figonzal.evaluatool.evalua1.modulo3.NivelesAdaptacionE1M3;
import cl.figonzal.evaluatool.evalua1.modulo4.ComprensionLectoraE1M4;
import cl.figonzal.evaluatool.evalua1.modulo4.ExactitudLectoraE1M4;
import cl.figonzal.evaluatool.evalua1.modulo4.ValoracionGlobalLecturaE1M4;
import cl.figonzal.evaluatool.evalua1.modulo5.OrtografiaFoneticaE1M5;
import cl.figonzal.evaluatool.evalua1.modulo5.OrtografiaVisualE1M5;
import cl.figonzal.evaluatool.evalua1.modulo6.CalculoNumeracionE1M6;
import cl.figonzal.evaluatool.interfaces.AbrirActivity;
import cl.figonzal.evaluatool.modelo.Evalua;
import io.github.luizgrp.sectionedrecyclerviewadapter.SectionAdapter;
import io.github.luizgrp.sectionedrecyclerviewadapter.SectionedRecyclerViewAdapter;

public class Evalua1Activity extends AppCompatActivity implements EvaluaAdapter.ClickListener, AbrirActivity {

    private FirebaseCrashlytics crashlytics;
    private SectionedRecyclerViewAdapter sectionedRecyclerViewAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_evalua1);
        crashlytics = FirebaseCrashlytics.getInstance();

        Toolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setTitleTextColor(getResources().getColor(R.color.colorOnPrimary, getTheme()));
        setSupportActionBar(toolbar);

        ActionBar actionBar = getSupportActionBar();
        assert actionBar != null;
        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setHomeAsUpIndicator(R.drawable.ic_arrow_back_white_24dp);
        actionBar.setTitle(getString(R.string.TOOLBAR_EVALUA_1));

        configurarExpandedList();
    }

    private void configurarExpandedList() {
        sectionedRecyclerViewAdapter = new SectionedRecyclerViewAdapter();

        //SUBTIMES - SUBMODULO 1
        List<Evalua> subItems1 = new ArrayList<>();
        subItems1.add(new Evalua(getString(R.string.EVALUA_1_M1_SI_1)));

        //Submodulo2
        List<Evalua> subItems2 = new ArrayList<>();
        subItems2.add(new Evalua(getString(R.string.EVALUA_1_M2_SI_1)));
        subItems2.add(new Evalua(getString(R.string.EVALUA_1_M2_SI_2)));
        subItems2.add(new Evalua(getString(R.string.EVALUA_1_M2_SI_3)));
        subItems2.add(new Evalua(getString(R.string.EVALUA_1_EVALUA_GLOBAL)));

        //Submodulo3
        List<Evalua> subitems3 = new ArrayList<>();
        subitems3.add(new Evalua(getString(R.string.EVALUA_1_M3_SI_1)));

        //Submodulo4
        List<Evalua> subItems4 = new ArrayList<>();
        subItems4.add(new Evalua(getString(R.string.EVALUA_1_M4_SI_1)));
        subItems4.add(new Evalua(getString(R.string.EVALUA_1_M4_SI_2)));
        subItems4.add(new Evalua(getString(R.string.EVALUA_1_EVALUA_GLOBAL)));

        List<Evalua> subItems5 = new ArrayList<>();
        subItems5.add(new Evalua(getString(R.string.EVALUA_1_M5_SI_1)));
        subItems5.add(new Evalua(getString(R.string.EVALUA_1_M5_SI_2)));

        //Submodulo6
        List<Evalua> subItems6 = new ArrayList<>();
        subItems6.add(new Evalua(getString(R.string.EVALUA_1_M6_SI_1)));

        //HEADERS
        sectionedRecyclerViewAdapter.addSection(new EvaluaAdapter(getString(R.string.EVALUA_1_MODULO_1), subItems1, this));
        sectionedRecyclerViewAdapter.addSection(new EvaluaAdapter(getString(R.string.EVALUA_1_MODULO_2), subItems2, this));
        sectionedRecyclerViewAdapter.addSection(new EvaluaAdapter(getString(R.string.EVALUA_1_MODULO_3), subitems3, this));
        sectionedRecyclerViewAdapter.addSection(new EvaluaAdapter(getString(R.string.EVALUA_1_MODULO_4), subItems4, this));
        sectionedRecyclerViewAdapter.addSection(new EvaluaAdapter(getString(R.string.EVALUA_1_MODULO_5), subItems5, this));
        sectionedRecyclerViewAdapter.addSection(new EvaluaAdapter(getString(R.string.EVALUA_1_MODULO_6), subItems6, this));

        final RecyclerView recyclerView = findViewById(R.id.recyclerview);
        recyclerView.setLayoutManager(new LinearLayoutManager(getApplicationContext()));
        recyclerView.setAdapter(sectionedRecyclerViewAdapter);
    }

    @Override
    public void onItemRootViewClicked(@NonNull String sectionTitle, int itemAdapterPosition) {

        //MOdulo 1
        if (getString(R.string.EVALUA_1_MODULO_1).equals(sectionTitle)) {
            if (sectionedRecyclerViewAdapter.getPositionInSection(itemAdapterPosition) == 0) {

                abrirActividad(
                        this,
                        MemoriaAtencionE1M1.class,
                        getString(R.string.SUB_ITEM_CLICK),
                        getString(R.string.CLICK_MEMORIA_ATENCION)
                );
            }
        }

        //Modulo 2
        else if (getString(R.string.EVALUA_1_MODULO_2).equals(sectionTitle)) {
            if (sectionedRecyclerViewAdapter.getPositionInSection(itemAdapterPosition) == 0) {

                abrirActividad(
                        this,
                        SeriesE1M2.class,
                        getString(R.string.SUB_ITEM_CLICK),
                        getString(R.string.CLICK_SERIES)
                );
            } else if (sectionedRecyclerViewAdapter.getPositionInSection(itemAdapterPosition) == 1) {

                abrirActividad(
                        this,
                        ClasificacionesE1M2.class,
                        getString(R.string.SUB_ITEM_CLICK),
                        getString(R.string.CLICK_CLASIFICACION)
                );
            } else if (sectionedRecyclerViewAdapter.getPositionInSection(itemAdapterPosition) == 2) {

                abrirActividad(
                        this,
                        OrganizacionPerceptivaE1M2.class,
                        getString(R.string.SUB_ITEM_CLICK),
                        getString(R.string.CLICK_ORG_PERCEPTIVA)
                );
            } else if (sectionedRecyclerViewAdapter.getPositionInSection(itemAdapterPosition) == 3) {

                abrirActividad(
                        this,
                        ValoracionGlobalBasesE1M2.class,
                        getString(R.string.SUB_ITEM_CLICK),
                        getString(R.string.CLICK_VALORACION_GLOBAL)
                );
            }
        }

        //Modulo 3
        else if (getString(R.string.EVALUA_1_MODULO_3).equals(sectionTitle)) {
            if (sectionedRecyclerViewAdapter.getPositionInSection(itemAdapterPosition) == 0) {

                abrirActividad(
                        this,
                        NivelesAdaptacionE1M3.class,
                        getString(R.string.SUB_ITEM_CLICK),
                        getString(R.string.CLICK_NIVELES_ADAPTACION)
                );
            }
        }

        //MOdulo 4
        else if (getString(R.string.EVALUA_1_MODULO_4).equals(sectionTitle)) {
            if (sectionedRecyclerViewAdapter.getPositionInSection(itemAdapterPosition) == 0) {

                abrirActividad(
                        this,
                        ComprensionLectoraE1M4.class,
                        getString(R.string.SUB_ITEM_CLICK),
                        getString(R.string.CLICK_COMPRENSION_LECTORA)
                );
            } else if (sectionedRecyclerViewAdapter.getPositionInSection(itemAdapterPosition) == 1) {
                abrirActividad(
                        this,
                        ExactitudLectoraE1M4.class,
                        getString(R.string.SUB_ITEM_CLICK),
                        getString(R.string.CLICK_EXACTITUD_LECTORA)
                );
            } else if (sectionedRecyclerViewAdapter.getPositionInSection(itemAdapterPosition) == 2) {
                abrirActividad(
                        this,
                        ValoracionGlobalLecturaE1M4.class,
                        getString(R.string.SUB_ITEM_CLICK),
                        getString(R.string.CLICK_VALORACION_GLOBAL)
                );
            }
        }

        //MOdulo 5
        else if (getString(R.string.EVALUA_1_MODULO_5).equals(sectionTitle)) {
            if (sectionedRecyclerViewAdapter.getPositionInSection(itemAdapterPosition) == 0) {

                abrirActividad(
                        this,
                        OrtografiaFoneticaE1M5.class,
                        getString(R.string.SUB_ITEM_CLICK),
                        getString(R.string.CLICK_ORTOGRAFIA_FONETICA)
                );
            } else if (sectionedRecyclerViewAdapter.getPositionInSection(itemAdapterPosition) == 1) {

                abrirActividad(
                        this,
                        OrtografiaVisualE1M5.class,
                        getString(R.string.SUB_ITEM_CLICK),
                        getString(R.string.CLICK_ORTOGRAFIA_VISUAL)
                );
            }
        }

        //Modulo6
        else if (getString(R.string.EVALUA_1_MODULO_6).equals(sectionTitle)) {
            if (sectionedRecyclerViewAdapter.getPositionInSection(itemAdapterPosition) == 0) {

                abrirActividad(
                        this,
                        CalculoNumeracionE1M6.class,
                        getString(R.string.SUB_ITEM_CLICK),
                        getString(R.string.CLICK_CAL_NUMERACION)
                );
            }
        }
    }

    @Override
    public void onHeaderRootViewClicked(@NonNull final EvaluaAdapter section) {

        SectionAdapter sectionAdapter = sectionedRecyclerViewAdapter.getAdapterForSection(section);

        final boolean wasExpanded = section.isExpanded();
        final int previousItemsTotal = section.getContentItemsTotal();

        section.setExpanded(!wasExpanded);
        sectionAdapter.notifyHeaderChanged();

        if (wasExpanded) {
            sectionAdapter.notifyItemRangeRemoved(0, previousItemsTotal);
        } else {
            sectionAdapter.notifyAllItemsInserted();
        }
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {

        if (item.getItemId() == android.R.id.home) {

            Log.d(getString(R.string.CLOSE_EVALUA_1), getString(R.string.ACTIVIDAD_CERRADA));

            crashlytics.log(getString(R.string.CLOSE_EVALUA_1) + getString(R.string.ACTIVIDAD_CERRADA));

            finish();

            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void abrirActividad(Activity activity, Class<?> calledActivity, String log_title, String log_reponse) {
        Log.d(log_title, log_reponse);
        crashlytics.log("D/" + log_title + ": " + log_reponse);

        Intent intent = new Intent(activity, calledActivity);
        startActivity(intent);
    }
}

File: app/src/main/java/cl/figonzal/evaluatool/evalua/evalua1/Evalua1Activity.kt
/*

 This file is subject to the terms and conditions defined in
 file 'LICENSE', which is part of this source code package

 Autor: Felipe González
 Email: felipe.gonzalezalarcon94@gmail.com

 Copyright (c) 2021

 Last modified 01-05-21 22:19
 */
package cl.figonzal.evaluatool.evalua.evalua1

import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import cl.figonzal.evaluatool.R
import cl.figonzal.evaluatool.adapter.EvaluaAdapter
import cl.figonzal.evaluatool.adapter.EvaluaAdapter.ClickListener
import cl.figonzal.evaluatool.databinding.ActivityEvalua1Binding
import cl.figonzal.evaluatool.modelo.Evalua
import cl.figonzal.evaluatool.utilidades.ConfigRoutes
import cl.figonzal.evaluatool.utilidades.RouteHandler
import cl.figonzal.evaluatool.utilidades.configActionBar
import cl.figonzal.evaluatool.utilidades.logInfo
import io.github.luizgrp.sectionedrecyclerviewadapter.SectionedRecyclerViewAdapter
import java.util.*

class Evalua1Activity : AppCompatActivity(), ClickListener {

    private lateinit var binding: ActivityEvalua1Binding
    private var sectionedRecyclerViewAdapter: SectionedRecyclerViewAdapter = SectionedRecyclerViewAdapter()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityEvalua1Binding.inflate(layoutInflater)
        setContentView(binding.root)

        configActionBar(R.string.EVALUA_1, binding.include.toolbar)
        configurarExpandedList()
    }

    private fun configurarExpandedList() {

        //SUBTIMES - SUBMODULO 1
        val subItems1 = listOf(
                Evalua(getString(R.string.EVALUA_1_M1_SI_1))
        )

        //Submodulo2
        val subItems2 = listOf(
                Evalua(getString(R.string.EVALUA_1_M2_SI_1)),
                Evalua(getString(R.string.EVALUA_1_M2_SI_2)),
                Evalua(getString(R.string.EVALUA_1_M2_SI_3)),
                Evalua(getString(R.string.EVALUA_1_EVALUA_GLOBAL))
        )

        //Submodulo3
        val subitems3 = listOf(
                Evalua(getString(R.string.EVALUA_1_M3_SI_1))
        )

        //Submodulo4
        val subItems4 = listOf(
                Evalua(getString(R.string.EVALUA_1_M4_SI_1)),
                Evalua(getString(R.string.EVALUA_1_M4_SI_2)),
                Evalua(getString(R.string.EVALUA_1_EVALUA_GLOBAL))
        )

        val subItems5 = listOf(
                Evalua(getString(R.string.EVALUA_1_M5_SI_1)),
                Evalua(getString(R.string.EVALUA_1_M5_SI_2))
        )

        //Submodulo6
        val subItems6 = listOf(
                Evalua(getString(R.string.EVALUA_1_M6_SI_1))
        )


        //HEADERS
        sectionedRecyclerViewAdapter.apply {
            addSection(EvaluaAdapter(getString(R.string.EVALUA_1_MODULO_1), subItems1, this@Evalua1Activity))
            addSection(EvaluaAdapter(getString(R.string.EVALUA_1_MODULO_2), subItems2, this@Evalua1Activity))
            addSection(EvaluaAdapter(getString(R.string.EVALUA_1_MODULO_3), subitems3, this@Evalua1Activity))
            addSection(EvaluaAdapter(getString(R.string.EVALUA_1_MODULO_4), subItems4, this@Evalua1Activity))
            addSection(EvaluaAdapter(getString(R.string.EVALUA_1_MODULO_5), subItems5, this@Evalua1Activity))
            addSection(EvaluaAdapter(getString(R.string.EVALUA_1_MODULO_6), subItems6, this@Evalua1Activity))
        }

        //Recycler View
        binding.rv1.apply {
            layoutManager = LinearLayoutManager(applicationContext)
            adapter = sectionedRecyclerViewAdapter
        }

    }

    override fun onItemRootViewClicked(sectionTitle: String, itemAdapterPosition: Int) {

        ConfigRoutes.routeMap[getString(R.string.routeMapEvalua1)]?.let {
            RouteHandler.handleRoutes(
                    it,
                    sectionTitle,
                    sectionedRecyclerViewAdapter,
                    itemAdapterPosition,
                    this
            )
        }
    }

    override fun onHeaderRootViewClicked(section: EvaluaAdapter) {

        val sectionAdapter = sectionedRecyclerViewAdapter.getAdapterForSection(section)
        val wasExpanded = section.isExpanded
        val previousItemsTotal = section.contentItemsTotal

        section.isExpanded = !wasExpanded
        sectionAdapter.notifyHeaderChanged()

        when {
            wasExpanded -> {
                sectionAdapter.notifyItemRangeRemoved(0, previousItemsTotal)
            }
            else -> {
                sectionAdapter.notifyAllItemsInserted()
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {

        if (item.itemId == android.R.id.home) {
            logInfo(R.string.ACTIVIDAD_CERRADA)
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }
}